system/lib/emmalloc.cpp

```diff
@@ -0,0 +1,1187 @@
+/*
+ * Simple minimalistic but efficient malloc/free.
+ *
+ * Assumptions:
+ *
+ *  - 32-bit system.
+ *  - Single-threaded.
```

Comment:
I had thought there might be a way at compile-time to get the value of the compiler's thread-model (e.g. "single" or "posix") so we could assert on that too, but I didn't find one. It might be worth adding that to clang for wasm, since we expect users to care more about the distinction that on other platforms.